Configure the console by default only on particular platforms

bgilbert opened the pull request:
Stop specifying `console=` kernel arguments by default, except on specific arch/platform pairs where we know we need them.  Do the same with console configuration in `grub.cfg`.  More info in coreos/fedora-coreos-tracker#567.

Read `platforms.yaml` from the config repo (https://github.com/coreos/fedora-coreos-config/pull/1181) to determine what console settings should be applied for each arch/platform.  Convert the table for the current arch to JSON and copy it to `/boot/coreos/platforms.json` in the image, since coreos-installer needs the same data when overriding the platform ID at install time (https://github.com/coreos/coreos-installer/pull/605).  If `platforms.yaml` is missing, continue applying the previous defaults.

We do this in two places: `create_disk.sh` configures console settings for `metal` and `qemu` images from `platforms.yaml`.  For other platforms, `gf-set-platform` (previously `gf-platformid`) reads `platforms.json` from the image, undoes any `qemu`-specific settings, and applies settings for the target platform.

cc @coreos/multi-arch

bgilbert commented:
/test all

bgilbert commented:
/test all

cgwalters commented:
> As a result of this change, we need a way to continue getting a serial console in kola run and kola qemuexec

We already have `kola qemuexec --kargs` which works by using libguestfs to edit a copy of the disk image before booting.

cgwalters commented:
> /test all

That only triggers Prow, not Jenkins.  I think CCI was restarted, and lost state.  Going to try the close/reopen.

dustymabe commented:
webhooks are failing to be delivered.. investigating

cgwalters reviewed (dismissed):
Overall LGTM

dustymabe commented on src/gf-platformid:
```diff
@@ -41,14 +41,35 @@ coreos_gf_run_mount ro "${tmp_dest}"
+fi
 coreos_gf upload "${tmpd}"/bls.conf "${blscfg_path}"
 
+# Inject grub.cfg commands
+if [ -n "$extra_grub_cmds" ] && [ "$(coreos_gf exists /boot/grub2/grub.cfg)" = "true" ]; then
+    coreos_gf download /boot/grub2/grub.cfg "${tmpd}"/grub.cfg
+    sed -i -e 's@^# CONSOLE-SETTINGS$@'"${extra_grub_cmds}"'@' "${tmpd}"/grub.cfg
+    coreos_gf upload "${tmpd}"/grub.cfg /boot/grub2/grub.cfg
+fi
+
```
does ppc64le use grub? just wondering if we need to conditionalize this?

bgilbert replied:
ppc64le doesn't use grub but petitboot reads the `grub.cfg`.  s390x doesn't use grub, but presumably won't define any `grub_commands`.  The only reason this block is conditionalized is so we don't try to inject the legacy grub commands on s390x.

dustymabe commented on src/grub.cfg:
```diff
@@ -49,9 +49,9 @@ function load_video {
-terminal_input serial console
-terminal_output serial console
+# Any non-default console settings will be substituted here.
+# CONSOLE-SETTINGS
```
optional: might be worth linking to this PR here (I know git-blame works, but sometimes code gets moved around).

bgilbert replied:
Eh, this is potentially a user-facing config file, and `CONSOLE-SETTINGS` is greppable.

dustymabe reviewed (approved):
LGTM - some questions/comments.

bgilbert commented:
The legacy path wasn't properly setting kargs or GRUB commands in the metal or qemu images, since those don't go through `gf-platformid`.  Refactored to fix that, and added a per-arch disable knob to aid migration per https://github.com/coreos/fedora-coreos-config/pull/1181#discussion_r700599359.

Tested the contents of `grub.cfg` and kargs, and the existence of `platforms.json`,  in these contexts:

- No `platforms.yaml` (qemu, metal, aws)
- Arch with `legacy: true` in `platforms.yaml` (qemu, metal, aws)
- Arch missing from `platforms.yaml` (qemu, metal, aws)
- Platform missing from `platforms.yaml` (aws)
- Platform with kargs + GRUB commands in `platforms.yaml` (aws)
- Platform with kargs only in `platforms.yaml` (aws)
- Platform with GRUB commands only in `platforms.yaml` (aws)

Ready for review.

dustymabe commented:
mostly LGTM - will try to test later today

dustymabe reviewed (dismissed):
LGTM - will do more testing when this hits the pipeline

Prashanth684 commented on src/create_disk.sh:
```diff
@@ -61,6 +65,17 @@ udevtrig() {
+# If the platforms table has the legacy flag set for this arch, pretend
+# the table doesn't exist
+if [ -n "$platforms" ] && jq -e ".$arch.legacy" < "$platforms" > /dev/null; then
+    platforms=
```
so the legacy flag will be added by default for ppc64le ? the console won't log anything without the console kargs on ppc64le

bgilbert replied:
The legacy flag is just for managing the migration to the new behavior, and will go away in some number of weeks.  In non-legacy mode, kola on ppc64le will continue to [inject](https://github.com/coreos/coreos-assembler/pull/2400/files#diff-110657a50e0db550de5910ee3d75e73e9cad0f5f990fb7c6b2cd341cdf1f4951R72) a console argument in `kola run` and `kola qemuexec`, but the images themselves will have no console argument by default.  The assumption is that users can configure a non-default console if they need it.  Are you saying that that behavior is not correct for ppc64le?

Prashanth684 replied:
from the OCP perspective, for baremetal UPI installs, this is totally fine, the console karg can be injected. i'm thinking more about the OCP IPI install scenario where the installer generates the ignition config - maybe for that we can inject console in the ignition through installer/MCO.

bgilbert replied:
We want a default console that works for most users, so it'd be helpful to better understand the scope of the problem.  How large is the mismatch between kernel defaults and what's needed?  E.g., if there are no console arguments, by default will the kernel never have a working console?  Have no working console on most hardware?  Create a functioning console on the video card, but ppc64le users always expect to use a serial console?

Prashanth684 replied:
`console=hvc0 console=tty0` is needed on qemu for a working console. for baremetal, i can double check but iirc from a previous conversation with some IBM folks a while ago it is needed for metal too, but I will confirm.

Prashanth684 replied:
after the testing done on baremetal, looks like this is only needed for qemu, so we are all set.

bgilbert replied:
Updated https://github.com/coreos/fedora-coreos-config/pull/1181 to apply `console=hvc0 console=tty0` on ppc64le qemu.  Thanks for the research!

cgwalters commented:
This is awesome work by the way!

> As a result of this change, we need a way to continue getting a serial console in kola run and kola qemuexec

You touched on this but the way I think of this, the problem really parallels the bare metal case of initial provisioning and wanting debug output, particularly for the Ignition run.   As well as for the initial ISO boot.  Or I guess, any time Ignition runs.

> We could embed the location of the BLS config near the partition table and use qemu-img map --output=json plus that custom header to inject kargs directly into the qcow2 file, similar to how we inject kargs into the ISO. 

Yeah, that's quite doable. 

But a downside of this is that if we e.g. land that code in coreos-assembler, it won't be useful for everyone using e.g. `virt-install` type flows.    Or at least, it'd need to be replicated there.

It's easy to think of `qemu` platform as not a production thing, but it's really immensely useful for both us *and* our users to e.g. validate Ignition configs before deploying them on a real platform.   Just a few days ago I played "FCOS user" and validated in qemu with `cosa run` an Ignition config I was creating for a little home server device.

This relates to https://github.com/coreos/fedora-coreos-tracker/issues/235 - basically us extracting the `cosa run` and all the sugar around that out into a separate project.

cgwalters commented:
One alternative path we could investigate for bare metal is in some cases, making it easy to "console everywhere we can" but then turn it off once the server is out of provisioning phase.    I commented on this elsewhere but in OCP we recently had a large customer escalation that was greatly exacerbated by `printk` performance issues on serial console.

So after this lands it'd be good to document that flow of adding the console kargs, but then removing them once Ignition is done.

bgilbert commented:
Updated to support configuring `metal` and `qemu` images in `platforms.yaml` as well.  This allows https://github.com/coreos/fedora-coreos-config/pull/1181 to continue enabling serial console on `qemu`, per OOB discussion.

Tested the contents of `grub.cfg` and kargs, and the existence of `platforms.json`, on:

- `metal` (`create_disk.sh` with no customizations)
- `qemu` (`create_disk.sh` with customizations)
- `aws` (`gf-set-platform` with customizations)
- `aliyun` (`gf-set-platform` with no customizations)

in these cases:

- No `platforms.yaml` (the legacy case)
- `platforms.yaml` from https://github.com/coreos/fedora-coreos-config/pull/1181
- Arch missing from `platforms.yaml`
- Platform without kargs in `qemu` image

Ready for another review.

dustymabe commented:
Do you know what the coupling of this looks like? i.e. Is it safe to land and since FCOS `stable` won't have a `platforms.yaml` it will just be a no-op for the build in 10 days?

bgilbert commented on src/gf-set-platform:
```diff
@@ -41,14 +41,43 @@ coreos_gf_run_mount ro "${tmp_dest}"
 # Remove any platformid currently there
 sed -i -e 's, ignition.platform.id=[a-zA-Z0-9]*,,g' "${tmpd}"/bls.conf
 sed -i -e '/^options / s,$, ignition.platform.id='"${platformid}"',' "${tmpd}"/bls.conf
+if [ -n "$remove_kargs" ]; then
+    # Remove existing qemu-specific kargs
+    sed -i -e '/^options / s@ '"${remove_kargs}"'@@' "${tmpd}"/bls.conf
+fi
```
Right.  The problem with such a mark is that it'd be user-visible in the machine's command line for the life of the node.  The advantage to the current approach is that it doesn't have any user-visible side effects.

dustymabe replied:
Fair.. Though we *could* push `$ignition_firstboot` (something that already exists that we are adding) to the end of the command line and let that be our mark.

I'm good with what we have here too. Just tossing ideas.

bgilbert replied:
Yeah, we could.  It'd give that variable a slightly non-obvious side effect, though.

I'm not too bothered by what we have, though, so I'm inclined to stick with it to avoid another rework.

bgilbert commented:
The idea is that we land this and https://github.com/coreos/coreos-installer/pull/605 first (in any order), then https://github.com/coreos/fedora-coreos-config/pull/1181 once we're ready.

bgilbert commented on src/cmd-buildextend-metal:
```diff
@@ -217,10 +209,17 @@ cat >image-dynamic.json << EOF
 EOF
 cat "${image_json}" image-dynamic.json | jq -s add > image-for-disk.json
+platforms_json=
+if [ -e "${configdir}/platforms.yaml" ]; then
+    platforms_json="${workdir}/tmp/platforms.json"
+    yaml2json "${configdir}/platforms.yaml" "${platforms_json}"
+fi
```
Yup, noted.  Let's not block on that, though.

bgilbert commented on src/create_disk.sh:
```diff
@@ -29,6 +29,8 @@ Options:
     --help: show this help
     --kargs: kernel CLI args
+    --platform: Ignition platform ID
+    --platforms: JSON-formatted platforms.yaml
```
Updated.

dustymabe reviewed (dismissed):
LGTM

jlebon commented on src/create_disk.sh:
```diff
@@ -406,7 +406,7 @@ s390x)
     # stage on s390x, either through zipl->grub2-emu or zipl standalone.
     # See https://github.com/coreos/ignition-dracut/issues/84
-    # A similar hack is present in https://github.com/coreos/coreos-assembler/blob/main/src/gf-platformid#L55
+    # A similar hack is present in https://github.com/coreos/coreos-assembler/blob/main/src/gf-set-platform#L55
```
Aside/optional: that line number will eventually become stale. We should probably either make it use a permalink, or just say `gf-set-platform`.

jlebon commented on src/create_disk.sh:
```diff
@@ -29,6 +29,8 @@ Options:
     --disk: disk device to use
     --help: show this help
     --kargs: kernel CLI args
+    --platform: Ignition platform ID
```
Minor: `--platform-id` would be more canonical

bgilbert replied:
Hmm, I'm not seeing any instances of `--platform-id` in the codebase.  `kola run` already has `--platform`, as does `cosa artifact-disk manual`.

jlebon replied:
Hmm right. I didn't mean literally `--platform-id` the switch, but platform ID the name. But indeed, we're a bit inconsistent about this between switches and documentation/internal variable names, etc...

jlebon commented on src/cmd-buildextend-metal:
```diff
@@ -217,10 +209,17 @@ cat >image-dynamic.json << EOF
             --config "$(pwd)"/image-for-disk.json \
             --kargs "\"${kargs}\"" \
+            --platform "${ignition_platform_id}" \
+            ${platforms_json:+--platforms-json "${platforms_json}"} \
```
I think it'd be cleaner to just add to `disk_args` in the if-statement above.

bgilbert replied:
Fair.  Once we're done ratcheting `platforms.yaml` into the config, this argument will be deconditionalized in a followup PR, so I'm inclined to leave as is to reduce churn.

bgilbert replied:
Legacy removal in https://github.com/coreos/coreos-assembler/pull/2899.

jlebon commented on src/create_disk.sh:
```diff
@@ -62,6 +68,19 @@ udevtrig() {
+    # Add legacy kargs and console settings
+    platform_grub_cmds='serial --speed=115200\nterminal_input serial console\nterminal_output serial console'
+    DEFAULT_TERMINAL=$(. $(dirname "$0")/cmdlib.sh; echo $DEFAULT_TERMINAL)
+    platform_kargs="console=tty0 console=${DEFAULT_TERMINAL},115200n8"
```
Previously, on s390x and aarch64 we would unset `tty` so that `DEFAULT_TERMINAL` wouldn't be injected. Seems like we're losing that here. Is that on purpose?

bgilbert replied:
It isn't.  Good catch!

jlebon commented on src/create_disk.sh:
```diff
@@ -364,7 +383,16 @@ install_grub_cfg() {
+        # Copy platforms table if it's non-empty for this arch
+        if jq -e ".$arch" < "$platforms_json" > /dev/null; then
+            mkdir -p "$rootfs/boot/coreos"
+            jq ".$arch" < "$platforms_json" > "$rootfs/boot/coreos/platforms.json"
```
Remind me again, why does the coreos-installer side require it in `/boot`? If it were only in `/usr`, could it not just mount the rootfs to find it? I know I initially was the one in favour of [only having it in `/boot`](https://github.com/coreos/fedora-coreos-config/pull/1181#issuecomment-905596476), but now with https://github.com/coreos/coreos-assembler/pull/2839, there is a push for shipping "image-building metadata" in the rootfs.

So, could we ship this not in `/boot` but instead in `/usr/lib/coreos-assembler/` alongside `image.json`? You said in another comment to not block on this, but not sure if you feel the same if a change in path was in the picture. Otherwise, I fear we'll indeed end up with it both in `/boot` and `/usr` (which obviously isn't the end of the world, but it'd be nice to have a single source of truth).

bgilbert replied:
The installer already mounts `/boot` and doesn't currently know or care about `/`.  We could change that, but the installer would also need to navigate the ostree repo, tying `install -p` to ostree.  To me that seems like a worse tradeoff than maintaining a private copy of `platforms.json` for use by cosa.

(Even if we moved `platforms.json` to `/usr`, it shouldn't go in `/usr/lib/coreos-assembler`, since that path strongly implies that it's a private internal API.  Maybe something like `/usr/share/coreos/platforms.json`?)

Note that the path change wouldn't be sufficient for #2839 compatibility, since we're currently injecting `platforms.json` during image build and not during ostree build.

jlebon replied:
Yeah, tricky. Seems doable still, but unsure about the tradeoff also.

cgwalters commented:
A lot going on here, and RHCOS stuff is in bugfix mode for OCP 4.11 freeze...I think we should cut a `rhcos-4.11` branch here now (potentially before https://github.com/coreos/coreos-assembler/pull/2811 too) and start using it now to de-risk landing this a bit more?

cgwalters commented:
OK actually, no reason to block on 4.11 branching, I just pushed https://github.com/coreos/coreos-assembler/tree/rhcos-4.11

cgwalters commented:
https://github.com/openshift/release/pull/29057

bgilbert commented:
@dustymabe, yup, I think we're good to go!

dustymabe commented:
Local testing testing of https://github.com/coreos/fedora-coreos-pipeline/pull/533 went well. I think this is now unblocked.

newkit commented:
Hi, is this planned to be part of an RHCOS release?

bgilbert commented:
@newkit Yes, that's the plan.

bgilbert commented:
There's no public commitment, but I'd expect it'd land soon after the functionality lands in FCOS.  Note that this PR is only part of it; the overall FCOS tracking bug is https://github.com/coreos/fedora-coreos-tracker/issues/567.
